### PR TITLE
Test suite improvements

### PR DESCRIPTION
**Pull request by wesbland**
Most tests are returning 0, even when they fail. This makes it difficult to use a particular test in something like a git bisection which relies on a non-zero return code to check correctness.

Potentially, this could also simplify the test scripts to just check the return value instead of the correct output of " No Errors".

Change order of arg parsing so the more general option doesn't gobble the more specific ones.

### CODE REVIEW

**Comment by wesbland** [comment]
test:jenkins/ch4/all

**Comment by raffenet** [comment]
test:jenkins/ch3/all

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp

**Comment by wesbland** [comment]
I don't know what to make of these osx no error checking builds. Is this my fault?

**Comment by wesbland** [comment]
I guess it's the same problem as https://github.com/pmodels/mpich/pull/2621. Forgot about that one.

**Comment by wesbland** [comment]
We found a few tests that were messed up because they were xfailed and I missed them the first time around. I think this is good to go now.

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
@wesbland several tests are failing on this PR.  Can you take a look?

**Comment by wesbland** [comment]
Should be better now. Let's see:

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
Still failing big time, @wesbland

**Comment by wesbland** [comment]
I'm working on it. Weird that this doesn't fail for ib64.

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: test/mpi/coll/allred.c [hunk 1]
```diff
@@ -18,7 +18,7 @@
 #endif
 
 int count, size, rank;
-int cerrcnt;
+int errs;
```
Comment (pavanbalaji) [inline test/mpi/coll/allred.c]: Can separate out the name change to a different commit?  It's unrelated to this.

File: test/mpi/coll/alltoallv.c [hunk 2]
```diff
@@ -29,10 +29,10 @@ int main(int argc, char **argv)
     int *sbuf, *rbuf;
     int rank, size;
     int *sendcounts, *recvcounts, *rdispls, *sdispls;
-    int i, j, *p, err;
+    int i, j, *p, errs;
```
Comment (pavanbalaji) [inline test/mpi/coll/alltoallv.c]: Name change should be a separate commit.

File: test/mpi/coll/alltoallv0.c [hunk 3]
```diff
@@ -27,11 +27,11 @@ int main(int argc, char **argv)
     int *sbuf, *rbuf;
     int rank, size;
     int *sendcounts, *recvcounts, *rdispls, *sdispls;
-    int i, *p, err;
+    int i, *p, errs;
```
Comment (pavanbalaji) [inline test/mpi/coll/alltoallv0.c]: Name change should be a separate commit.

File: test/mpi/coll/bcasttest.c [hunk 4]
```diff
@@ -19,7 +19,7 @@ int main(int argc, char **argv)
     int i, rank, reps, n;
     int bVerify = 1;
     int sizes[NUM_SIZES] = { 100, 64 * 1024, 128 * 1024, 1024 * 1024 };
-    int num_errors = 0;
+    int errs = 0;
```
Comment (pavanbalaji) [inline test/mpi/coll/bcasttest.c]: Name change should be a separate commit.

File: test/mpi/coll/gather_big.c [hunk 5]
```diff
@@ -56,7 +56,7 @@ int main(int argc, char *argv[])
 
     MPI_Gather(sendbuf, COUNT, MPI_LONG, recvbuf, COUNT, MPI_LONG, ROOT, MPI_COMM_WORLD);
 
-    int lerr = 0;
+    int errs = 0;
```
Comment (pavanbalaji) [inline test/mpi/coll/gather_big.c]: Name change should be a separate commit.

Comment (pavanbalaji) [reply test/mpi/coll/gather_big.c]: I gave up at this point.

**Comment by wesbland** [comment]
I ended up cleaning a bunch of other stuff that was bugging me. So I have no expectations that these tests will pass cleanly.

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
This should be better now.

The `psend` failures should be xfailed as a replacement for the `allpairf_psend` failures.

test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
Ok @pavanbalaji / @raffenet - The tests are finally looking right now. You can look again.

**Comment by pavanbalaji** [comment]
@wesbland a bunch of the RMA tests are failing.

**Comment by wesbland** [comment]
Dang it. Why didn’t those fail before...

**Comment by wesbland** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
The tests look good.  I've signed off on most of the commits.  I've added a fixup commit.

I didn't signoff on two of the commits: `mtest: Have all tests return non-0 on failure` and `mtest: Remove unnecessary prints and reductions`.  These use `MTest` routines even when the program does not initialize with `MTest_Init`.  Perhaps there needs to be a separate patch that uses `MTest_Init` and `MTest_Finalize` consistently in all tests.

**Comment by wesbland** [comment]
I started off by adding `MTest_Init` to match with `MTest_Finalize`, but I stopped because while `MTest_Init` calls `MPI_Init` if necessary, `MTest_Finalize` does not. So it didn't seem necessary to do both.

If you want both, that's fine, but I'd prefer to see `MTest_Finalize` also call `MPI_Finalize` if it hasn't been called already.

I'll add a patch to do that.

**Comment by wesbland** [comment]
Do you want me to push the patches you've already signed off?

**Comment by pavanbalaji** [comment]
Whether MPI_Init should be replaced when MTest_Init is an orthogonal issue.  I agree that makes sense, but my comment was only that MTest_Init/Finalize should be used together and other MTest functionality shouldn’t be used without them.

**Comment by wesbland** [comment]
I added some patches that:

1. Add `MTest_Init` where `MTest_Finalize` is used
2. Add `MPI_Finalize` to `MTest_Finalize`
3. Remove `MPI_Init` and `MPI_Finalize` usage

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp

**Comment by wesbland** [comment]
test:jenkins/ch3/mxm

Get a clang build too

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
I have no idea what's going on with `threaded_sr`. I've reverted all of the meaningful changes to that file and it doesn't fail in any systems where I can replicate the problem.

**Comment by wesbland** [comment]
I'm not able to reproduce this problem on my MacOS laptop. In fact, for me this test fails even with `master`. Maybe it should be xfailed?

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
Finally got all the checks done. Hit the button quick @pavanbalaji.

**Comment by pavanbalaji** [comment]
@wesbland sorry, I was on travel and just got back.  This branch seems to have some conflicts.  Can you fix it so I can review it?

**Comment by wesbland** [comment]
@pavanbalaji - Should be good to go again.

**Comment by pavanbalaji** [comment]
@wesbland I think my comments were somewhat misunderstood in these patches.  Here's what I was expecting:

1. Cleanup patch to call `MTest_Init` and `MTest_Finalize` in all tests (and removing the unnecessary print statements).

2. Use `MTestReturnValue` in all tests that would be on top of the previous patch.

3. Convert all tests to use just `MTest_Init` and `MTest_Finalize` instead of `MPI_Init` and `MPI_Finalize`.

I think this patch is incorrect: `mtest: Initialize test with MTest_Init`.  It is doing both `MPI_Init` and `MTest_Init`.  I agree that they should be unified, and that's the bullet number 3 above, but that's a separate thing.

**Comment by wesbland** [comment]
I’m not sure I understand what you mean is different than this in the current branch. 

Anyway, I’m out of the office at least all this week (just occasional access to my laptop) so I won’t be able to do much about it until at least Monday.

**Comment by wesbland** [comment]
So @pavanbalaji I'm not sure how what I have here is different from what you're suggesting in your comment above. The only difference is the ordering. I've shuffled `mtest: Rename files to remove hyphens` to move before the patches you describe below so the "prereqs" are bundled together. I've also renamed a few patches to make them more clear.

> 1. Cleanup patch to call `MTest_Init` and `MTest_Finalize` in all tests (and removing the unnecessary print statements).

This is done with these patches:

```
* mtest: Use MTest_Finalize whenever possible
* mtest: Initialize test with MTest_Init
```

> 2. Use `MTestReturnValue` in all tests that would be on top of the previous patch.

This is done here:

```
* mtest: Have all tests use MTestReturnValue
```

While it can be moved up, it's a bunch of conflict resolutions for nothing in my mind.

> 3. Convert all tests to use just `MTest_Init` and `MTest_Finalize` instead of `MPI_Init` and `MPI_Finalize`.

This is here:

```
* mtest: Change MTest_Finalize to call MPI_Finalize
* mtest: Remove extra use of MPI_init/finalize
```

These _could_ be combined into a single patch, but the first change is important and I think should be called out as its own patch.

**Comment by pavanbalaji** [comment]
I have two primary issues with this branch:

1. Some of the intermediate patches are using `MTest` routines without initializing the tests with `MTest_Init`.  I think that can be fixed by squashing those three patches.

2. One of the patches is introducing both `MPI_Init` and `MTest_Init` into the same file.  I think that's incorrect.  I guess squashing the last commit will fix this too.

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:intel:ch4/ofi/sockets/all/force-am-disabled/shm-exclusive,shm-no/debug

**Comment by wesbland** [comment]
test:intel:ch4/ofi/sockets/all/force-am-disabled/shm-exclusive,shm-no/debug

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
I fixed the build issues after your changes @pavanbalaji. Do you want to rebase and signoff?

**Comment by pavanbalaji** [comment]
@wesbland can you squash my fixup commit (if you agree) and signoff on my patch?

**Comment by wesbland** [comment]
@pavanbalaji - Done.
